react-common/components/controls/TeachingBubble.tsx

```diff
@@ -0,0 +1,291 @@
+    const finishLabel = lf("Finish");
+    const hasPrevious = stepNumber > 1;
+    const hasNext = stepNumber < totalSteps;
+    const hasSteps = totalSteps > 1;
+
+    const classes = classList(
+        "teaching-bubble-container",
+        className
+    );
+
+    return ReactDOM.createPortal(<FocusTrap className={classes} onEscape={onClose}>
+        <div className="teaching-bubble-cutout"></div>
+        <div className="teaching-bubble-arrow"></div>
+        <div id={id}
+            className="teaching-bubble"
+            role={role || "dialog"}
+            aria-hidden={ariaHidden}
+            aria-label={ariaLabel}
+            aria-describedby={ariaDescribedBy}
+            aria-labelledby="teaching-bubble-title">
+            <Button
+                className="teaching-bubble-close"
+                onClick={onClose}
+                title={closeLabel}
+                ariaLabel={closeLabel}
+                rightIcon="fas fa-times-circle"
+            />
+            <div className="teaching-bubble-content">
+                <strong>{targetContent.title}</strong>
+                <p>{targetContent.description}</p>
+                {hasSteps && <div className="teaching-bubble-footer">
+                    {stepNumber && totalSteps && <div className="teaching-bubble-steps">
+                        {stepNumber} of {totalSteps}
+                    </div>}
+                    <div className="teaching-bubble-navigation">
+                        {hasPrevious && <Button
+                            className="secondary-button"
+                            onClick={onBack}
+                            title={backLabel}
+                            ariaLabel={backLabel}
+                            label={backLabel}
+                        />}
+                        {hasNext && <Button
+                            className="primary-button"
+                            onClick={onNext}
+                            title={nextLabel}
+                            ariaLabel={nextLabel}
+                            label={nextLabel}
+                        />}
+                        {!hasNext && <Button
+                            className="primary-button"
+                            onClick={onClose}
+                            title={finishLabel}
+                            ariaLabel={finishLabel}
+                            label={finishLabel}
+                        />}
+                    </div>
+                </div>}
+                {(!hasSteps) && <div className="teaching-bubble-footer">
+                    <Button
+                        className="primary-button"
+                        onClick={onClose}
+                        title={confirmLabel}
+                        ariaLabel={confirmLabel}
+                        label={confirmLabel}
+                    />
+                </div>}
+            </div>
+        </div>
+    </FocusTrap>, parentElement || document.body)
+}
```

Comment:
Oh, I also wanted to note here, and this may be more work than what it's worth. There's a lot of functionality in here that directly relates to the bubble arrow itself. Something that we may want to to do make the logic in `TeachingBubble.tsx` more simple is to break the bubble arrow into a separate component and then pass in, as a prop, which way the arrow is facing, and then do the computation from there.

Reply:
Yea, I actually thought about making the cutout a separate component. I could also make the bubble arrow its own component. Not sure if I'll do it for this PR, but might be worth doing in the future. Thanks for the suggestion.